vignettes/case-study-scaling.Rmd

```diff
@@ -0,0 +1,222 @@
+
+## Benchmark
+
+The first step is benchmarking, and we'll do it in several sub-steps:
+
+1. Record a realistic session using `shinyloadtest::record_session`
+1. Run `shinycannon` with 1 user, producing a directory of "baseline" metric data
+1. Pick a number **n**, the number of users we want to be able to use the app at the same time
+1. Run `shinycannon` with **n** users, producing another directory of metric data
+
+### Record a session
+
+First you should install the `shinyloadtest` R package if you haven't yet:
+
+```
+devtools::install_github('rstudio/shinyloadtest')
+```
+
+Next, obtain the "solo" URL of the deployed app in RStudio Connect. You can do this by navigating to the application from the list of published content, clicking the circular button at the top right of the screen, and then clicking "Open Solo". This will open a new window. Copy the URL from this new window.
+
+![](../docs/inst/img/rsc_open_solo.png)
+
+Back in the R console, run `record_session` with the solo URL as an argument:
+
+```
+library(shinyloadtest)
+record_session("https://rsc.example.com/content/1234/")
+```
+
+This starts a local web server listening on http://127.0.0.1:8600/ and should also open a new browser window to this URL. This local web server is a proxy that relays content to and from your RStudio Connect instance, recording interactions so that they can be played back later. It's in this window that you should model what you imagine would be a typical user interaction. Click around, change things, and observe new outputs.
+
+Do this for 1-2 minutes. When you're done, close the browser tab. You should notice the function return in the R console.
+
+On disk, in the directory where you started R, will be a new file: `recording.log`. There may also be other files like `recording.log.post.0`. These complementary files are created when you upload files or if you use an htmlwidget like [DT](https://github.com/rstudio/DT) that uses HTTP POST to send data.
+
+Together, these files constitute a **recording**. They are the input to `shinycannon`.
+
+### Baseline run
+
+Now that you've made a recording, you're ready for your first `shinycannon` run. For the first run, you should just test with 1 user. This will generate metrics analogous to those you experienced when you made your recording, and will serve as a baseline for subsequent runs.
+
+> Note: It's best to run shinycannon on a machine close to the target machine on the network. We don't want RStudio Connect and shinycannon to contend for resources on the same machine. You should also ensure the machine you run shinycannon on has enough memory and bandwidth. shinycannon requires about 512kb of memory per worker, so (allowing for OS memory usage) a machine with 4GB of RAM should support in the ballpark of 5000-8000 worker threads.
```

Comment:
Hm, that's a good point, but I think it probably depends on the nature of the recording. We can definitely get away with way more workers than cores on a machine because playback is dominated by IO waiting. I can imagine a command-line tool that analyzes a recording and produces some guesstimates about the hardware necessary to achieve various worker counts.

Maybe for this document we should just put in a note about the particular hardware we've used for our successful tests?